backend/plugins/authorization.js

```diff
@@ -0,0 +1,87 @@
+import {contains} from 'ramda';
+import {getSetting} from '../settings.js';
+import {generateAndSaveAccessToken, isElectron, getAccessTokenExpiry} from '../utils/authUtils';
+import Logger from '../logger';
+import fetch from 'node-fetch';
+
+const ESCAPED_ROUTES = [
+  new RegExp('^/$'),
+  new RegExp('^/login$'),
+  new RegExp('^/static/'),
+  new RegExp('^/oauth2$')
+]
+
+function accessTokenIsValid(access_token) {
+    return getSetting('ACCESS_TOKEN') === access_token;
+}
+
+export function PlotlyOAuth() {
+
+    function isAuthorized(req, res, next) {
+        const path = req.href();
+
+        if (!getSetting('AUTH_ENABLED')) {
+          return (next());
+        }
+        // Auth is disabled for certain urls:
+        if (ESCAPED_ROUTES.some(path.match.bind(path))) {
+          return (next());
+        }
+
+        // No Auth for electron apps:
+        if (isElectron()) {
+          return (next());
+        }
+        if (!accessTokenIsValid(req.cookies['db-connector-auth-token'])) {
+
+            if (!req.cookies['plotly-auth-token']) {
+              res.json(401, {error: {message: 'Please login to access this page.'}});
+              return;
+            }
+
+            const plotlyAuthToken = req.cookies['plotly-auth-token'];
+
+            fetch(`${getSetting('PLOTLY_API_URL')}/v2/users/current`, {
+                headers: {'Authorization': `Bearer ${plotlyAuthToken}`}
+            })
+            .then(userRes => userRes.json().then(userMeta => {
+
+              if (userRes.status !== 200) {
+                res.json(401, {error: {message: 'Please login to access this page.'}});
+                return;
+              }
+              else {
+                  if (!contains(userMeta.username, getSetting('ALLOWED_USERS'))) {
+
+                      // Remove any existing credentials and return error
+                      res.clearCookie('db-connector-auth-token');
+                      res.clearCookie('plotly-auth-token');
+                      res.json(403, {error: {message: `User ${userMeta.username} is not allowed to view this app`}})
+                      return;
+
+                  } else {
+
+                      const dbConnectorAccessToken = generateAndSaveAccessToken();
+                      res.setCookie('db-connector-auth-token',
+                                    dbConnectorAccessToken, {
+                                        'maxAge': getAccessTokenExpiry(),
+                                        'path': '/'
```

Comment:
I think we should make `secure` a setting like `SSL_ENABLED` - this is similar to how the dash app manager works: https://github.com/plotly/dokku-integration/blob/master/manager/backend/server.py#L99-L105
And I still think that the more specific the path, the better. This will allow users to run this behind a proxy and reduce the thread of a cookie leaking through an XSS in their main app code. Also similar to how the app manager works: https://github.com/plotly/dokku-integration/blob/master/manager/backend/server.py#L103